File: rust/hangman/src/hangman.rs
//! The state and the rules of a single hangman round.

/// The number of life points a player starts with.
pub const MAX_LIFE_POINTS: u32 = 10;

/// The character that is shown instead of a letter that is still hidden.
const HIDDEN_MARKER: char = '_';

/// The outcome of a single guess.
///
/// An `enum` lists all possible outcomes. Because `match` has to cover every
/// variant, the compiler makes it impossible to forget one of the cases.
#[derive(Debug, PartialEq, Eq)]
pub enum Guess {
    /// The letter occurs in the word, at least one position was revealed.
    Correct,
    /// The letter does not occur in the word, one life point is lost.
    Wrong,
    /// The letter was already guessed before, so nothing happens.
    AlreadyUsed,
}

/// Holds everything that belongs to one round of the game.
pub struct Hangman {
    /// The secret word, split into single characters.
    word: Vec<char>,
    /// One flag per character of the word: `true` if it is already revealed.
    revealed: Vec<bool>,
    /// Every letter that was guessed so far, sorted alphabetically.
    used_letters: Vec<char>,
    /// The remaining life points.
    life_points: u32,
}

impl Hangman {
    /// Creates a new round for the given word.
    ///
    /// # Arguments
    ///
    /// * `word` - The word to guess. It has to contain at least one letter.
    ///
    /// # Returns
    ///
    /// * `Ok(hangman)` - A round that is ready to be played.
    /// * `Err(message)` - The word contains no letter at all.
    pub fn new(word: &str) -> Result<Self, &'static str> {
        // A `String` cannot be indexed by position, because it is UTF-8
        // encoded. A `Vec<char>` can, which is needed to reveal single letters.
        let word: Vec<char> = word.chars().collect();

        // `any` returns true as soon as one character satisfies the condition.
        if !word.iter().any(|character| character.is_alphabetic()) {
            return Err("The word has to contain at least one letter!");
        }

        // Characters that are no ASCII letters (like spaces, punctuation or digits)
        // can never be guessed. They are revealed right from the start,
        // otherwise such a word could never be completed.
        let revealed = word
            .iter()
            .map(|character| !character.is_ascii_alphabetic())
            .collect();

        Ok(Self {
            word,
            revealed,
            used_letters: Vec::new(),
            life_points: MAX_LIFE_POINTS,
        })
    }

    /// Returns the number of characters of the word.
    pub fn word_length(&self) -> usize {
        self.word.len()
    }

    /// Returns the remaining life points.
    ///
    /// Rust naming conventions omit a `get_` prefix for such accessors.
    pub fn life_points(&self) -> u32 {
        self.life_points
    }

    /// Returns the secret word.
    pub fn word(&self) -> String {
        self.word.iter().collect()
    }

    /// Guesses a letter and updates the state of the round.
    ///
    /// # Arguments
    ///
    /// * `letter` - The guessed letter in upper case.
    ///
    /// # Returns
    ///
    /// What the guess did: [`Guess::Correct`], [`Guess::Wrong`] or
    /// [`Guess::AlreadyUsed`].
    pub fn guess(&mut self, letter: char) -> Guess {
        // A letter may only be used once, so a repeated guess is free.
        if self.used_letters.contains(&letter) {
            return Guess::AlreadyUsed;
        }

        // Remember the letter and keeps the list sorted for a tidy output.
        self.used_letters.push(letter);
        self.used_letters.sort_unstable();

        // Reveal every position where the letter occurs. A letter can appear
        // more than once, so the whole word is checked.
        let mut found = false;
        for (index, &character) in self.word.iter().enumerate() {
            if character == letter {
                self.revealed[index] = true;
                found = true;
            }
        }

        if found {
            Guess::Correct
        } else {
            // `saturating_sub` stops at zero. A plain `- 1` on an unsigned
            // integer would panic if the value was already zero.
            self.life_points = self.life_points.saturating_sub(1);
            Guess::Wrong
        }
    }

    /// Returns `true` if every letter of the word is revealed.
    pub fn is_word_complete(&self) -> bool {
        // `all` returns true only if the condition holds for every element.
        self.revealed.iter().all(|&revealed| revealed)
    }

    /// Returns `true` if there are no life points left.
    pub fn is_player_dead(&self) -> bool {
        self.life_points == 0
    }

    /// Returns the word with hidden letters, for example `_ _ A _ _`.
    pub fn formatted_word(&self) -> String {
        let mut output = String::new();

        // `zip` walk both vectors in parallel and yield pairs.
        for (index, (&character, &revealed)) in self.word.iter().zip(&self.revealed).enumerate() {
            // Separate the characters with a space, but not before the first.
            if index > 0 {
                output.push(' ');
            }

            output.push(if revealed { character } else { HIDDEN_MARKER });
        }

        output
    }

    /// Returns the used letters separated by commas, for example `A,L`.
    pub fn formatted_used_letters(&self) -> String {
        // `join` needs strings, so every character becomes a short `String`.
        self.used_letters
            .iter()
            .map(|letter| letter.to_string())
            .collect::<Vec<String>>()
            .join(",")
    }
}

// Unit tests are only compiled when `cargo test` is run.
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_new_round_hides_every_letter() {
        let hangman = Hangman::new("HASE").unwrap();

        assert_eq!(hangman.formatted_word(), "_ _ _ _");
        assert_eq!(hangman.formatted_used_letters(), "");
        assert_eq!(hangman.life_points(), MAX_LIFE_POINTS);
        assert!(!hangman.is_word_complete());
        assert!(!hangman.is_player_dead());
    }

    #[test]
    fn a_correct_guess_reveals_every_occurrence() {
        let mut hangman = Hangman::new("HANGAR").unwrap();

        assert_eq!(hangman.guess('A'), Guess::Correct);
        assert_eq!(hangman.formatted_word(), "_ A _ _ A _");
        // A correct guess must not cost a life point.
        assert_eq!(hangman.life_points(), MAX_LIFE_POINTS);
    }

    #[test]
    fn a_wrong_guess_costs_one_life_point() {
        let mut hangman = Hangman::new("HASE").unwrap();

        assert_eq!(hangman.guess('X'), Guess::Wrong);
        assert_eq!(hangman.life_points(), MAX_LIFE_POINTS - 1);
        assert_eq!(hangman.formatted_word(), "_ _ _ _");
    }

    #[test]
    fn a_repeated_guess_changes_nothing() {
        let mut hangman = Hangman::new("HASE").unwrap();
        hangman.guess('X');

        assert_eq!(hangman.guess('X'), Guess::AlreadyUsed);
        // Still only one life point lost and one letter used.
        assert_eq!(hangman.life_points(), MAX_LIFE_POINTS - 1);
        assert_eq!(hangman.formatted_used_letters(), "X");
    }

    #[test]
    fn used_letters_are_sorted() {
        let mut hangman = Hangman::new("HASE").unwrap();

        hangman.guess('S');
        hangman.guess('A');
        hangman.guess('X');

        assert_eq!(hangman.formatted_used_letters(), "A,S,X");
    }

    #[test]
    fn a_word_is_complete_when_all_letters_are_found() {
        let mut hangman = Hangman::new("AB").unwrap();

        hangman.guess('A');
        assert!(!hangman.is_word_complete());
        hangman.guess('B');
        assert!(hangman.is_word_complete());
    }

    #[test]
    fn the_player_dies_after_ten_wrong_guesses() {
        let mut hangman = Hangman::new("A").unwrap();

        for letter in ['B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K'] {
            assert_eq!(hangman.guess(letter), Guess::Wrong);
        }

        assert!(hangman.is_player_dead());
        assert_eq!(hangman.life_points(), 0);
    }

    #[test]
    fn life_points_never_drop_below_zero() {
        let mut hangman = Hangman::new("A").unwrap();

        // Twelve wrong guesses, but only ten life points.
        for letter in ['B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'] {
            hangman.guess(letter);
        }

        assert_eq!(hangman.life_points(), 0);
    }

    #[test]
    fn a_word_without_letters_is_rejected() {
        assert!(Hangman::new("").is_err());
        assert!(Hangman::new("   ").is_err());
        assert!(Hangman::new("42").is_err());
    }

    #[test]
    fn a_word_with_non_ascii_letters_is_working() {
        // Chars like ` Ü()` cannot be typed by the player,
        // so they are revealed from the start and do not count as letters.
        let mut hangman = Hangman::new("AB CD (SÜSS)").unwrap();

        assert_eq!(hangman.formatted_word(), "_ _   _ _   ( _ Ü _ _ )");

        for letter in ['A', 'B', 'C', 'D', 'S'] {
            assert_eq!(hangman.guess(letter), Guess::Correct);
        }

        assert!(hangman.is_word_complete());
        assert_eq!(hangman.formatted_word(), "A B   C D   ( S Ü S S )");
        assert_eq!(hangman.life_points(), MAX_LIFE_POINTS);
    }
}

File: lua/hangman/hangman.lua
--- The number of life points a player starts with.
local MAX_LIVES = 10

--- Results returned by `Hangman:guess()`.
---@enum GuessResult
local GuessResult = {
  --- The letter occurs in the word and at least one position was revealed.
  CORRECT = 1,
  --- The letter does not occur in the word and one life is lost.
  WRONG = 2,
  --- The letter was already guessed before, so nothing happens.
  DUPLICATE = 3,
  --- The input was invalid, e.g. more than one character or not a valid letter.
  INVALID = 4,
}

--- Converts a string to upper case and turns 'ß' into 'SS'.
--- Lua's `string.upper` does not convert 'ß' like Python or Rust do.
---@param str string
---@return string
local function to_upper(str)
  str = str:gsub('ß', 'SS')
  return string.upper(str)
end

--- Converts user input into a valid uppercase letter for the game.
--- Whitespace is removed and invalid input returns `nil`.
---@param input string The user input.
---@return string? letter A single uppercase letter from A to Z or `nil` if invalid.
local function normalize_letter(input)
  -- Remove surrounding whitespace.
  input = input:match('^%s*(.-)%s*$')

  -- The input must contain exactly one UTF-8 character.
  -- For example, `#'ä'` returns 2 because it counts bytes,
  -- while `utf8.len('ä')` returns 1 because it counts characters.
  if utf8.len(input) ~= 1 then return nil end

  -- The German eszett 'ß' is converted to 'SS' when uppercased.
  -- It is treated as a single accepted input and mapped to 'S'.
  if input == 'ß' then return 'S' end

  -- Accept letters A-Z in any case and return them uppercase.
  if input:match('^[a-zA-Z]$') then return string.upper(input) end

  -- Reject all other characters.
  return nil
end

--- A single letter of the word and its reveal state.
---@class HangmanLetter
---@field char string The character.
---@field revealed boolean Whether the character is visible.

--- Represents a Hangman game instance.
---@class Hangman
---@field _letters HangmanLetter[] The letters of the word to guess.
---@field _used_letters table<string, boolean> Letters that have already been guessed.
---@field _lives integer Remaining lives.
local Hangman = {}
Hangman.__index = Hangman

--- Creates a new Hangman game with the given word.
--- Returns `nil` if the word is empty or not provided.
---@param word string The word to guess.
---@return Hangman? game The new Hangman instance or `nil` if the word is invalid.
---@return string? err Error message if creation failed.
function Hangman:new(word)
  if not word or word == '' then return nil, 'Word must not be empty.' end

  word = to_upper(word)

  local obj = setmetatable({}, self)

  obj._letters = {}

  -- Lua strings are byte sequences, so iterating with `gmatch('.')`
  -- does not handle UTF-8 characters correctly. `utf8.codes()` iterates
  -- over Unicode code points, allowing each character to be stored separately.
  for _, codepoint in utf8.codes(word) do
    local char = utf8.char(codepoint)

    table.insert(obj._letters, {
      char = char,
      revealed = not char:match('^[A-Z]$'),
    })
  end

  obj._used_letters = {}
  obj._lives = MAX_LIVES

  return obj
end

--- Returns the secret word.
---@return string word The secret word.
function Hangman:word()
  local chars = {}

  for _, letter in ipairs(self._letters) do
    table.insert(chars, letter.char)
  end

  return table.concat(chars)
end

--- Returns the number of characters of the word.
---@return integer length The word length.
function Hangman:word_length() return #self._letters end

--- Returns the current and maximum number of lives.
---@return integer current Current remaining lives.
---@return integer max Maximum possible lives.
function Hangman:get_lives() return self._lives, MAX_LIVES end

--- Processes a guessed letter and updates the current game state.
--- Returns whether the guess was invalid, correct, wrong, or already used
---@param letter string The guessed letter.
---@return GuessResult result The result of the guess.
function Hangman:guess(letter)
  local normalized_letter = normalize_letter(letter)
  if not normalized_letter then return GuessResult.INVALID end

  -- A letter may only be used once, so a repeated guess is free.
  if self._used_letters[normalized_letter] then return GuessResult.DUPLICATE end

  -- Remember the letter and keep the list sorted.
  self._used_letters[normalized_letter] = true

  -- Reveal every position where the letter occurs. A letter can appear
  -- more than once, so the whole word is checked.
  local found = false
  for _, word_letter in ipairs(self._letters) do
    if word_letter.char == normalized_letter then
      word_letter.revealed = true
      found = true
    end
  end

  if found then return GuessResult.CORRECT end

  -- If the letter was not found, one life point is lost.
  self._lives = self._lives - 1

  return GuessResult.WRONG
end

--- Checks whether all letters of the word have been revealed.
---@return boolean complete `true` if the word is fully revealed, otherwise `false`.
function Hangman:is_word_complete()
  for _, word_letter in ipairs(self._letters) do
    if not word_letter.revealed then return false end
  end
  return true
end

--- Checks whether the player has no remaining life points.
---@return boolean dead `true` if the player has no life points left, otherwise `false`.
function Hangman:is_player_dead() return self._lives <= 0 end

--- Returns the word formatted for display with unrevealed letters hidden.
--- Example: `_ P P _ _` for the word `APPLE` if only `P` is revealed.
---@return string formatted The formatted word.
function Hangman:display_word()
  local letters = {}

  for _, letter in ipairs(self._letters) do
    table.insert(letters, letter.revealed and letter.char or '_')
  end

  return table.concat(letters, ' ')
end

--- Returns the used letters formatted for display.
---@return string formatted The formatted used letters.
function Hangman:display_used_letters()
  local letters = {}

  for letter, _ in pairs(self._used_letters) do
    table.insert(letters, letter)
  end

  table.sort(letters)

  return table.concat(letters, ',')
end

Hangman.GuessResult = GuessResult

return Hangman
